domain/watcher.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package domain
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/eventsource"
+)
+
+type WatchableDBFactory = func() (changestream.WatchableDB, error)
+
+type WatcherFactory struct {
+	mu sync.Mutex
+
+	getDB       WatchableDBFactory
+	watchableDB changestream.WatchableDB
+	logger      eventsource.Logger
+}
+
+func NewWatcherFactory(WatchableDBFactory WatchableDBFactory, logger eventsource.Logger) *WatcherFactory {
+	return &WatcherFactory{
+		getDB:  WatchableDBFactory,
+		logger: logger,
+	}
+}
+
+func (f *WatcherFactory) newBaseWatcher() (*eventsource.BaseWatcher, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	var err error
+
+	if f.watchableDB == nil {
+		if f.watchableDB, err = f.getDB(); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
```

Comment:
Delete this and just call getDB every time. It's a cheap operation once it's been called once.

```suggestion
	watchableDB, err := f.getDB()
	if err != nil {
		return nil, errors.Trace(err)
	}
```

Reply:
No. `Runner.Worker` takes a lock, so all requests are synchronised. It is better to hold the reference once acquired.